### PR TITLE
ipnsfs

### PR DESCRIPTION
**Pull request by whyrusleeping**
In this PR i implement a 'mock' in memory mutable filesystem to be used as the backend for the fuse ipns interface.

This is PR number 6 of the ipns set

### CODE REVIEW

File: ipnsfs/dir.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+	fs        *Filesystem
+	parent    childCloser
+	childDirs map[string]*Directory
+	files     map[string]*file
+
+	node *dag.Node
+	name string
+	lock sync.Mutex
+
+	ref     int
```
Comment (cryptix) [inline ipnsfs/dir.go]: reference? to what?

File: ipnsfs/dir.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+type Directory struct {
+	fs        *Filesystem
+	parent    childCloser
+	childDirs map[string]*Directory
+	files     map[string]*file
+
+	node *dag.Node
+	name string
+	lock sync.Mutex
```
Comment (cryptix) [inline ipnsfs/dir.go]: i like the style of having the `mu` above what it is protecting.

Also we have `lock` and `refLock` here. maybe a more descriptove for `lock`? like `nodeMu` and `refMu` ?

File: ipnsfs/dir.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	dir, err := d.childDir(tpath[0])
+	if err != nil {
+		return nil, err
+	}
+	return dir.Open(tpath[1:], mode)
+}
+
+// consider combining into a single method...
```
Comment (cryptix) [inline ipnsfs/dir.go]: update comment?

File: ipnsfs/dir.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+		d.lock.Unlock()
+		return err
+	}
+
+	err = d.node.AddNodeLinkClean(name, nd)
+	if err != nil {
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
```
Comment (cryptix) [inline ipnsfs/dir.go]: `// unlock here otherwise the recursion will block` ?

File: ipnsfs/dir.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+	ufspb "github.com/jbenet/go-ipfs/unixfs/pb"
+)
+
+var ErrNotYetImplemented = errors.New("not yet implemented")
+var ErrInvalidChild = errors.New("invalid child node")
+
+type Directory struct {
+	fs        *Filesystem
+	parent    childCloser
+	childDirs map[string]*Directory
```
Comment (cryptix) [inline ipnsfs/dir.go]: These maps are fine without a Mutex?

File: ipnsfs/dir.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+	_, err := d.childDir(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("directory by that name already exists")
+	}
+	_, err = d.childFile(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("file by that name already exists")
```
Comment (cryptix) [inline ipnsfs/dir.go]: maybe reuse `os.ErrExist` ?

File: ipnsfs/dir.go (hunk 5 above)
Comment (cryptix) [inline ipnsfs/dir.go]: I see `lock` being used [here](https://github.com/jbenet/go-ipfs/pull/923/files#diff-85be37b82b3ffbe8d91c875595d79e07R239) to protected these maps as well. See my comment below.

File: ipnsfs/dir.go (hunk 1 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: `oxDEADC0DE`

File: ipnsfs/dir.go (hunk 4 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: im actually stripping out all the locking, it shouldnt be needed. Fuse properly serializes requests

File: ipnsfs/dir.go [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
+
+	return d.parent.closeChild(d.name, d.node)
+}
+
+// RenameEntry renames the child by 'oldname' of this directory to 'newname'
+func (d *Directory) RenameEntry(oldname, newname string) error {
```
Comment (cryptix) [inline ipnsfs/dir.go]: is `newname == ""` allowed?

File: ipnsfs/dir.go [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+		err := d.node.RemoveNodeLink(oldname)
+		if err != nil {
+			return err
+		}
+		err = d.node.AddNodeLinkClean(newname, dir.node)
+		if err != nil {
+			return err
+		}
+
+		delete(d.childDirs, oldname)
```
Comment (cryptix) [inline ipnsfs/dir.go]: not locked?

File: ipnsfs/dir.go [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+	}
+
+	err = d.node.AddNodeLinkClean(name, nd)
+	if err != nil {
+		return err
+	}
+
+	switch pbn.GetType() {
+	case ft.TDirectory:
+		d.childDirs[name] = NewDirectory(name, nd, d, d.fs)
```
Comment (cryptix) [inline ipnsfs/dir.go]: lock?

File: ipnsfs/file.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	dag "github.com/jbenet/go-ipfs/merkledag"
+	mod "github.com/jbenet/go-ipfs/unixfs/mod"
+
+	context "github.com/jbenet/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+)
+
+type File interface {
+	io.ReadWriteCloser
+	io.WriterAt
+	Seek(int64, int) (int64, error)
```
Comment (cryptix) [inline ipnsfs/file.go]: Argument names? i.e. `Seek(offset int64, whence int) (ret int64, err error)`

File: ipnsfs/file.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	Truncate(int64) error
+	FSNode
+}
+
+type file struct {
+	parent childCloser
+	fs     *Filesystem
+
+	name       string
+	hasChanges bool
```
Comment (cryptix) [inline ipnsfs/file.go]: Do you have a `write>seek back>read` test? `hasChanges` seems racy at first thought?

File: core/core.go [hunk 12]
```diff
@@ -288,6 +301,13 @@ func (n *IpfsNode) teardown() error {
 		}
 	}
```
Comment (jbenet) [inline core/core.go]: we need to remove `addCloser` no?

File: ipnsfs/dir.go (hunk 1 above)
Comment (jbenet) [inline ipnsfs/dir.go]: let's remove it

File: ipnsfs/dir.go (hunk 2 above)
Comment (jbenet) [inline ipnsfs/dir.go]: +1 to mutex above. and let's keep the locks on this -- im paranoid about screwing up user's data.

File: ipnsfs/dir.go (hunk 4 above)
Comment (jbenet) [inline ipnsfs/dir.go]: how certain are you of this? i think we should lock to be safe. the overhead is not a huge price to pay for safety of users' data when we're not eeking out every last second yet (we have way bigger problems to address). last thing we want is surprise consistency issues.

Comment (jbenet) [inline ipnsfs/dir.go]: @cryptix i think the lock must be unlocked before closeChild. 

one pattern i like for doing the defer in cases like these is:

```go
func() {
  lk.Lock()
  defer lk.Unlock()
  
  .....
}()

closeChild(...)
```

File: ipnsfs/dir.go [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+	}
+}
+
+// Open opens a file at the given path 'tpath'
+func (d *Directory) Open(tpath []string, mode int) (*File, error) {
+	if len(tpath) == 0 {
+		return nil, ErrIsDirectory
+	}
+	if len(tpath) == 1 {
+		fi, err := d.childFile(tpath[0])
```
Comment (jbenet) [inline ipnsfs/dir.go]: what if the last component happens to be a child directory? seems to me that the error isn't handled?

this could happen if I:

```
mkdir foo
mkdir foo/bar
open foo/bar
```

File: ipnsfs/dir.go [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	// search dag
+	for _, lnk := range d.node.Links {
+		if lnk.Name == name {
+			nd, err := lnk.GetNode(d.fs.dserv)
+			if err != nil {
+				return nil, err
+			}
+			i, err := ft.FromBytes(nd.Data)
```
Comment (jbenet) [inline ipnsfs/dir.go]: what if it isn't unixfs object? just error out?

File: ipnsfs/dir.go [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+		return nil, errors.New("file by that name already exists")
+	}
+
+	ndir := &dag.Node{Data: ft.FolderPBData()}
+	err = d.node.AddNodeLinkClean(name, ndir)
+	if err != nil {
+		d.lock.Unlock()
+		return nil, err
+	}
+	d.lock.Unlock()
```
Comment (jbenet) [inline ipnsfs/dir.go]: locking here is a bit sketch. do we have to unlock for `closeChild`? if so, maybe using func defer?

```
func() {
  d.lock.Lock()
  defer d.lock.Unlock()
  
  ... stuff
}()
```

File: ipnsfs/dir.go (hunk 13 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: then `ErrNoSuch` gets returned. Opening directories is disallowed

File: ipnsfs/dir.go (hunk 14 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: Yeah, it *should* error out. Although with protobuf, it might just parse whatever object it is as a unixfs struct. We honestly have no way of telling...

File: ipnsfs/dir.go (hunk 13 above)
Comment (jbenet) [inline ipnsfs/dir.go]: @whyrusleeping then why return `ErrIsDirectory` when `len(tpath) == 0` ? There's a disparity here.

Comment (jbenet) [inline ipnsfs/dir.go]: And it will still be tried to create (will call `NewFile`). I suggest adding:

```go
if err == <error that indicates dir> {
  return nil, ErrIsDirectory
} 
```
here

File: ipnsfs/dir.go [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+		}
+
+		err = d.node.AddNodeLinkClean(newname, nd)
+		if err != nil {
+			return err
+		}
+
+		delete(d.childDirs, oldname)
+		d.files[newname] = fi
+		return d.parent.closeChild(d.name, d.node)
```
Comment (jbenet) [inline ipnsfs/dir.go]: there's a lot of common code in this section and the one above. if someone comes in and changes the semantics of one, they may fail to change the other. the whole inner part of these ifs is the same except one line

File: ipnsfs/file.go [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+	return &File{
+		fs:     fs,
+		parent: parent,
+		name:   name,
+		mod:    dmod,
+	}, nil
+}
+
+func (fi *File) Write(b []byte) (int, error) {
+	fi.hasChanges = true
```
Comment (jbenet) [inline ipnsfs/file.go]: needs lock

File: ipnsfs/dir.go (hunk 13 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: Good point

File: ipnsfs/file.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+func (fi *File) Read(b []byte) (int, error) {
+	fi.lock.Lock()
+	defer fi.lock.Unlock()
+	return fi.mod.Read(b)
+}
+
+func (fi *File) Close() error {
+	fi.lock.Lock()
+	defer fi.lock.Unlock()
+	if fi.hasChanges {
```
Comment (jbenet) [inline ipnsfs/file.go]: this is a minor nit, but in general prefer a pattern like:

```go
if !fi.hasChanges {
  return nil
}

... the rest
```
than
```go

if fi.hasChanges {
  ... the rest
}
return nil
```

in the latter, there's more going on + more surface area for something to go wrong

File: ipnsfs/file.go [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+		}
+
+		nd, err := fi.mod.GetNode()
+		if err != nil {
+			return err
+		}
+
+		fi.lock.Unlock()
+		err = fi.parent.closeChild(fi.name, nd)
+		fi.lock.Lock()
```
Comment (jbenet) [inline ipnsfs/file.go]: is there a cleaner way to do this? a bit sketch.

Comment (jbenet) [inline ipnsfs/file.go]: also, not clear to the reader why it must be unlocked.

File: ipnsfs/system.go [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+		pkh, err := k.GetPublic().Hash()
+		if err != nil {
+			return nil, err
+		}
+
+		root, err := fs.NewKeyRoot(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		roots[u.Key(pkh).Pretty()] = root
```
Comment (jbenet) [inline ipnsfs/system.go]: does it have to `.Pretty()`? its more expensive than without.

File: ipnsfs/system.go [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+	if !ok {
+		return nil, ErrNoSuch
+	}
+
+	return r.Open(pathelem[1:], mode)
+}
+
+func (fs *Filesystem) Close() error {
+	for _, r := range fs.roots {
+		err := r.Publish(context.TODO())
```
Comment (jbenet) [inline ipnsfs/system.go]: might be nice for these all to happen concurrently.

File: ipnsfs/system.go [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+	// A pointer to the filesystem to access components
+	fs *Filesystem
+
+	// val represents the node pointed to by this key. It can either be a File or a Directory
+	val FSNode
+
+	repub *Republisher
+}
+
+func (fs *Filesystem) NewKeyRoot(parent context.Context, k ci.PrivKey) (*KeyRoot, error) {
```
Comment (jbenet) [inline ipnsfs/system.go]: doc? why would someone call this? 

We need to start documenting all the code. we should start today. we should all use `govet` and `golint`. maybe even put them as part of the test suite...

File: ipnsfs/system.go [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+
+	root := new(KeyRoot)
+	root.key = k
+	root.fs = fs
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	pointsTo, err := fs.nsys.Resolve(ctx, name)
+	if err != nil {
```
Comment (jbenet) [inline ipnsfs/system.go]: what are the error semantics here? what if routing is offline?

File: ipnsfs/system.go [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+func (kr *KeyRoot) GetValue() FSNode {
+	return kr.val
+}
+
+func (kr *KeyRoot) Open(tpath []string, mode int) (*File, error) {
+	if kr.val == nil {
+		// No entry here... what should we do?
+		panic("nyi")
```
Comment (jbenet) [inline ipnsfs/system.go]: What state might we be in if we lang here?

File: ipnsfs/system.go [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+	}
+	if len(tpath) > 0 {
+		// Make sure our root is a directory
+		dir, ok := kr.val.(*Directory)
+		if !ok {
+			return nil, fmt.Errorf("no such file or directory: %s", tpath[0])
```
Comment (jbenet) [inline ipnsfs/system.go]: why not use `ErrNoSuch`? seems like broken semantics to me

File: ipnsfs/system.go [hunk 26]
```diff
@@ -0,0 +1,291 @@
+package ipnsfs
```
Comment (jbenet) [inline ipnsfs/system.go]: this package needs a doc explaining its design. i dont readily understand what's going on, why a keyroot is needed, what the semantics are, etc.

File: ipnsfs/system.go [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+	}
+
+	child.Lock()
+	k, err := kr.fs.dserv.Add(nd)
+	if err != nil {
+		child.Unlock()
+		return err
+	}
+	child.Unlock()
+	// Dont want to hold the lock while we publish
```
Comment (jbenet) [inline ipnsfs/system.go]: why not?

File: ipnsfs/system.go [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+	if !ok {
+		return errors.New("child of key root not valid type")
+	}
+
+	nd, err := child.GetNode()
+	if err != nil {
+		return err
+	}
+
+	child.Lock()
```
Comment (jbenet) [inline ipnsfs/system.go]: why do we need to hold this lock?

File: ipnsfs/system.go [hunk 29]
```diff
@@ -0,0 +1,291 @@
...
+		Publish:      make(chan struct{}, 1),
+		root:         root,
+	}
+}
+
+func (np *Republisher) Touch() {
+	select {
+	case np.Publish <- struct{}{}:
+	default:
+	}
```
Comment (jbenet) [inline ipnsfs/system.go]: is it ok for the message to be lost entirely? what if i change something as a publish is happening, this touch is missed and i no longer modify anything?

File: namesys/publisher.go [hunk 30]
```diff
@@ -136,3 +137,31 @@ func ValidateIpnsRecord(k u.Key, val []byte) error {
 	}
 	return nil
 }
+
+// InitializeKeyspace sets the ipns record for the given key to
+// point to an empty directory.
+// TODO: this doesnt feel like it belongs here
```
Comment (jbenet) [inline namesys/publisher.go]: where to move it?

File: pin/pin.go [hunk 31]
```diff
@@ -200,6 +200,8 @@ func (p *pinner) IsPinned(key util.Key) bool {
 }
 
 func (p *pinner) RemovePinWithMode(key util.Key, mode PinMode) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
```
Comment (jbenet) [inline pin/pin.go]: i think things like this should be their own PR

File: ipnsfs/dir.go (hunk 16 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: just gonna remove this method. Fuse does renames in a way that makes it impossible to use this method

File: ipnsfs/dir.go (hunk 13 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: So, i was about to go in and fix this up when i realized that fuse doesnt even use this, so this method (and all other `Open` methods) could be removed. The only caller for it is in the small test i have in `ipnsfs/system_test.go`

File: ipnsfs/system.go (hunk 20 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: Yes. thats the link name. we cant expect users to type the raw bytes of the key:
```
$ cd /ipns/\u0231\u4532....
```

File: ipnsfs/system.go (hunk 21 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: Hrm, are you okay throwing the errors away then?

File: ipnsfs/system.go (hunk 22 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: probably should unexport this. users shouldnt call this.

File: ipnsfs/system.go (hunk 23 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: resolution can still occur offline, it just checks the local datastore for the entry. If it fails, then we dont have an entry for the node. Every time a publish occurs, its stored in the local datastore.

File: ipnsfs/system.go (hunk 24 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: This state means that somehow a `KeyRoot` was constructed, but no value was assigned. It can only be the result of programmer error.

I can keep the panic, and change the text to be a little more informative for any future meddler in this code, or i can remove it, and make other things panic if a KeyRoot is made incorrectly.

File: ipnsfs/system.go (hunk 27 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: because holding a lock while waiting on network IO will cause things to hang. Plus, it doesnt really matter if we are publishing multiple changes simultaneously. There are some funky conditions we can get into with one taking longer than the one following it, but that will be solved later on when we do some sort of versioning in ipns

File: ipnsfs/system.go (hunk 29 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: Its not missed, the only time a `Touch` will not send, is if there has already been another touch since the last publish finished. Its essentially raising a flag saying 'Hey! we need an update'. It doesnt matter if someone else comes along to tell us we need an update, we already know (the channel is buffered)

File: namesys/publisher.go (hunk 30 above)
Comment (whyrusleeping) [inline namesys/publisher.go]: I'm really not sure... I put it three different places before it ended up here, and it still doesnt feel quite right.

File: ipnsfs/system.go (hunk 20 above)
Comment (jbenet) [inline ipnsfs/system.go]: i mean that it could be converted elsewhere (on interfaces instead of everywhere) but it's fine.

File: ipnsfs/system.go (hunk 21 above)
Comment (jbenet) [inline ipnsfs/system.go]: could probably merge the errors into one, or something.

Comment (jbenet) [inline ipnsfs/system.go]: just worried about how long this will take, given routing delay

File: ipnsfs/system.go (hunk 22 above)
Comment (jbenet) [inline ipnsfs/system.go]: agreed. though still-- doc it, as it's not to the reader what's going on.

File: ipnsfs/system.go (hunk 24 above)
Comment (jbenet) [inline ipnsfs/system.go]: > change the text to be a little more informative for any future meddler in this code

:+1:

File: ipnsfs/system.go (hunk 27 above)
Comment (jbenet) [inline ipnsfs/system.go]: i meant, document it there for others to be aware of the decision tradeoffs

File: ipnsfs/dir.go (hunk 9 above)
Comment (whyrusleeping) [inline ipnsfs/dir.go]: locked around the entire function

**Comment by jbenet** [comment]
test failures:
1. https://build.protocol-dev.com/job/network-test/2006/console
2. https://build.protocol-dev.com/job/ipfs_test_all_commits_on_branch/1134/console 

(1) not sure why it would be failing from this PR
(2) looks like the daemon problem back to haunt us. cc @cryptix

**Comment by cryptix** [comment]
re (2): it looks like something else is running on `:5001`..?

```
expecting success: 
  IPFS_PID=$! &&
  pollEndpoint -ep=/version -v -tout=1s -tries=60 2>poll_apierr > poll_apiout ||
  test_fsh cat actual_daemon || test_fsh cat daemon_err || test_fsh cat poll_apierr || test_fsh cat poll_apiout

> cat actual_daemon
Initializing daemon...
initializing ipfs node at /private/var/lib/jenkins/workspace/ipfs_test_all_commits_on_branch/src/github.com/jbenet/go-ipfs/test/sharness/trash directory.t0060-daemon.sh/.go-ipfs
generating 4096-bit RSA keypair...done
peer identity: QmasuPgG7vwy5AMq8MuygsgKS6GHohKu4GpQLKRrmRwFsf
to get started, enter:

	ipfs cat /ipfs/QmV9E5oNhFDWeRV8NQ91G7vp3pvff2S4UYkn89T5koSzqo/readme

API server listening on /ip4/127.0.0.1/tcp/5001
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080

> cat daemon_err
[0;37m00:24:18.512 [31mERROR [0;34m  cmd/ipfs: [0mlisten tcp 127.0.0.1:8080: bind: address already in use [0;37m<autogenerated>:24[0m
Error: listen tcp 127.0.0.1:5001: bind: address already in use

> cat poll_apierr

> cat poll_apiout
time="2015-03-16T00:24:10Z" level=debug msg=starting timeout=1s tries=60 url="http://127.0.0.1:5001/version" when=2015-03-16 00:24:10.121066804 +0000 UTC 
time="2015-03-16T00:24:10Z" level=debug msg="get failed" error="Response not OK. 404 404 Not Found" tries=60 
time="2015-03-16T00:24:11Z" level=debug msg="get failed" error="Response not OK. 404 404 Not Found" tries=59 
...
```

**Comment by jbenet** [comment]
Interesting! wonder what it could be-- whether it's a clash with some other proc

**Comment by whyrusleeping** [comment]
neither test failure seems to be remotely related to my changes... bleh

**Comment by jbenet** [comment]
@whyrusleeping network-test failed again: https://build.protocol-dev.com/job/network-test/2012/console `Error: blockstore: block not found`.

the common failures of network-test are: 
- it actually succeeds but gets marked as failure (not sure why)
- it runs too long on a starved VM and gets killed early.

other than that network-test is reliable. this failure is not common for network-test. something is going on. either in this PR, or in code recently merged.

**Comment by whyrusleeping** [comment]
@jbenet it also failed over in @chriscool's latest PR

**Comment by jbenet** [comment]
Oh uh. This is bad news. Master may be broken. We should bisect 



—

**Comment by jbenet** [comment]
> @jbenet it also failed over in @chriscool's latest PR

No, the failure in #931 is: https://build.protocol-dev.com/job/network-test/2009/console which is the "succeeded but mistakenly marked as failed."

Ok, i've tested master 10x times (on jupiter) and it **failed 0/10 times**.
I've tested this PR 5x times (on jupiter) and it **failed 5/5 times**.

There's something wrong in this PR. @whyrusleeping can you not repro the failure locally? Look at the test console output: https://build.protocol-dev.com/job/network-test/2012/console

**Comment by whyrusleeping** [comment]
Thats really weird, it looks like only client_1 ever starts doing anything..
